File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> elements;
int length;
int value;
int tg;

void search(int depth, int total)
{
    
    if(depth>= length)
    {
        if(total == tg)
        {
            value +=1;
        }
        return;
    }
    else
    {
        for(int i=0; i<2;i++)
        {
            if(i==0)
            {
                total += elements[depth];
                search(depth +1, total);
                total -= elements[depth];
            }
            else
            {
                total -= elements[depth];
                search(depth +1, total);
                total += elements[depth];
            }
        }
    }
}

int solution(vector<int> numbers, int target) {
    int answer = 0;
    
    tg = target;
    
    length = numbers.size();
    
    for(int i=0; i<length; i++)
    {
        elements.push_back(numbers[i]);
    }
    
    search(0, 0);
    
    answer = value;
    
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.js
let total = 0;

function dfs(numbers,  target, sum, idx) {
    if(idx === numbers.length ) {
        if(sum === target) total +=1;
        return;
    }
    
    dfs(numbers, target, sum + numbers[idx], idx+1);
    dfs(numbers, target, sum - numbers[idx], idx+1);
}

function solution(numbers, target) {
    var answer = 0;
    
    dfs(numbers, target, -numbers[0] , 1);
    dfs(numbers, target, numbers[0] , 1);
    
    answer = total;
    
    return answer;
}